scipy/special/_cosine.c

```diff
@@ -0,0 +1,292 @@
+                       1.3728570152788793e-05,
+                       -0.0008916919927321117,
+                       0.022927496105281435,
+                       -0.25287619213750784,
+                       1.0};
+    double v;
+
+    v = polevl(t, coeffs, sizeof(coeffs) / sizeof(coeffs[0]) - 1);
+    return v;
+}
+
+
+//
+// Part of the asymptotic expansion of the inverse function at p=0.
+//
+// See, for example, the wikipedia article "Kepler's equation"
+// (https://en.wikipedia.org/wiki/Kepler%27s_equation).  In particular, see the
+// series expansion for the inverse Kepler equation when the eccentricity e is 1.
+//
+static
+double _poly_approx(double s)
+{
+    double s2;
+    double p;
+    double px;
+    double coeffs[] = {1.1911667949082915e-08,
+                       1.683039183039183e-07,
+                       43.0/17248000,
+                       1.0/25200,
+                       1.0/1400,
+                       1.0/60,
+                       1.0};
+    //
+    // p(s) = s + (1/60) * s**3 + (1/1400) * s**5 + (1/25200) * s**7 +
+    //        (43/17248000) * s**9 + (1213/7207200000) * s**11 +
+    //        (151439/12713500800000) * s**13 + ...
+    //
+    // Here we include terms up to s**13.
+    //
+    s2 = s*s;
+    p = s*polevl(s2, coeffs, sizeof(coeffs)/sizeof(coeffs[0]) - 1);
+    return p;
+}
+
+
+//
+// cosine distribution inverse CDF (aka percent point function).
+//
+double cosine_invcdf(double p)
+{
+    double x;
+    double y, y2;
+    double f0, f1, f2;
+    int i;
+    int sgn = 1;
+    double prevx;
+
+    if ((p < 0) || (p > 1)) {
+        return NAN;
+    }
+    if (p <= 1e-48) {
+        return -M_PI;
+    }
+    if (p == 1) {
+        return M_PI;
+    }
+
+    if (p > 0.5) {
+        p = 1.0 - p;
+        sgn = -1;
+    }
+
+    if (p < 0.0925) {
+        x = _poly_approx(cbrt(12*M_PI*p)) - M_PI;
+    }
+    else {
+        y = M_PI*(2*p - 1);
+        y2 = y*y;
+        x = y * _p2(y2) / _q2(y2);
+    }
+
+    // For p < 0.0018, the asymptotic expansion at p=0 is sufficently
+    // accurate that no more work is needed.  Similarly, for p > 0.42,
+    // the Pade approximant is sufficiently accurate.  In between these
+    // bounds, we refine the estimate with Halley's method.
+
+    if ((0.0018 < p) && (p < 0.42)) {
+        // Several iterations of Halley's method:
+        //    f(x) = pi + x + sin(x) - y,  f'(x) = 1 + cos(x), f''(x) = -sin(x)
+        // where y = 2*pi*p.
+        prevx = x;
+        for (i = 0; i < 12; ++i) {
+            f0 = M_PI + x + sin(x) - 2*M_PI*p;
+            f1 = 1 + cos(x);
+            f2 = -sin(x);
+            x = x - 2*f0*f1/(2*f1*f1 - f0*f2);
+            if (fabs(x - prevx)/x < 5e-15) {
+                break;
+            }
+            prevx = x;
+        }
+        if (i == 12) {
+            // This should not happen!  Extensive testing shows that the loop
+            // should break before completing 12 iterations.
```

Comment:
The idea is that you'd prefer users to get NaN (and maybe report a bug) than to potentially get an inaccurate value?

Reply:
Yes.  However, I did some more testing, and found that on a grid of 10 million points in the interval [0.0018, 0.42], only one step of Halley's method was needed, so I removed the loop and the check.  Now it just does the single Halley step.